File: chapter02/smallStepSemantic/While.rb
class While < Struct.new(:condition, :body)
  def reduce(environment)
    [If.new(condition, Sequence.new(body, self), DoNothing.new), environment]
  end

  def reducible?
    true
  end

  def to_s
    "while (#{condition}) {#{body}}"
  end

  def inspect
    "#{self}"
  end
end

File: chapter02/BigStepSemantic/While.py
from chapter02.BigStepSemantic import Boolean


class While(object):
    def __init__(self, condition, body):
        self.condition = condition
        self.body = body

    def evaluate(self, environment):
        if self.condition.evaluate(environment).value == Boolean(True).value:
            return self.evaluate(self.body.evaluate(environment))
        elif self.condition.evaluate(environment).value == Boolean(False).value:
            return environment
